### ISSUE TITLE
Support Type Hints

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/pygmt#2794]
*This issue is still in draft mode. Type hints are something new to me, so I need to learn more and will keep improving the issue description.*

**References:**

- https://docs.python.org/3/library/typing.html
- https://typing.readthedocs.io/en/latest/source/libraries.html
- https://typing.readthedocs.io/en/latest/source/best_practices.html

**Why we should support type hints**

1. Better auto-completion for parameter values #2793

**Comment by weiji14** [comment]
Personally I do use type hints (and tried to add it to `pygmt.grdtrack` in 2019 apparently - https://github.com/GenericMappingTools/pygmt/pull/308#discussion_r287089810), but it can be a lot of work to add them across the whole project (and a bit tough for new contributors to learn).

Maybe we can start small with just `typing.Literal` since that's a practical one for when there's a choice of str type arguments, e.g. `load_earth_relief(...,  data_source: Literal["igpp", "gebco", "gebcosi", "synbath"], ...)` in #2793? Other types like bool/int/float don't offer any autocomplete so may be less useful for now.

**Comment by seisman** [comment]
Yes, it's definitely a lot of work and a long-term goal. We need to split it into multiple smaller jobs so that people can help.

**Comment by seisman** [comment]
What are the best type hints for parameters that expect a fixed number of values?

In GMT, many parameters accept a sequence with a fixed number of elements. For example, the `panel` parameter can take a sequence of two values, such as `panel=(1, 2)`.

There are two common options for type hints:

1. `tuple[int, int]`: It clearly indicates that the parameter expects exactly two integers. However, it restricts the argument to a tuple, excluding lists or NumPy arrays.
2. `Sequence[int]`: It accepts any sequence-like object (e.g., tuple, list, array), but does not explicitly specify the expected number of elements. [We're doing the checking at run time in the AliasSystem class.]

Currently, both styles are used, but I prefer option 2 for flexibility and plan to update the code for consistency. 

@GenericMappingTools/pygmt-maintainers Comments?
